### PR TITLE
Flash from url

### PR DESCRIPTION
**Pull request by thundron**

### CODE REVIEW

**Comment by nazrhom** [comment]
@balena-ci retest

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 1]
```diff
@@ -73,36 +77,158 @@ function getState() {
...
+const URLSelector = ({ done }: { done: (imageURL: string) => void }) => {
+	const [imageURL, setImageURL] = React.useState('');
+	const [recentImages, setRecentImages]: [
+		string[],
+		(value: React.SetStateAction<string[]>) => void,
+	] = React.useState([]);
+	React.useEffect(() => {
+		const fetchRecentUrlImages = async () => {
+			try {
+				const recentUrlImages: any = await jsonStorage.get('recentUrlImages');
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:89]: Please avoid `any`.

Comment (zvin) [reply lib/gui/app/components/source-selector/source-selector.tsx:89]: Also provide a default empty list or the code below will break.

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 2]
```diff
@@ -73,36 +77,158 @@ function getState() {
...
+				console.error(err);
+			}
+		};
+		fetchRecentUrlImages();
+	}, []);
+	return (
+		<Modal
+			done={async () => {
+				const sanitizedRecentUrls = _.uniq(
+					_.reject([...recentImages, imageURL], _.isEmpty),
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:101]: This will not work if `recentImages` is `undefined`.

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 3]
```diff
@@ -73,36 +77,158 @@ function getState() {
...
+	const [recentImages, setRecentImages]: [
+		string[],
+		(value: React.SetStateAction<string[]>) => void,
+	] = React.useState([]);
+	React.useEffect(() => {
+		const fetchRecentUrlImages = async () => {
+			try {
+				const recentUrlImages: any = await jsonStorage.get('recentUrlImages');
+				setRecentImages(recentUrlImages);
+			} catch (err) {
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:91]: Don't use a try catch here, this has to work.

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 4]
```diff
@@ -73,36 +77,158 @@ function getState() {
...
+};
+
+interface Flow {
+	icon: any;
+	onClick: (evt: MouseEvent) => void;
+	label: string;
+}
+
+const FlowSelector = styled(
+	({ flow, ...props }: { flow: Flow; props?: any }) => {
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:154]: Please avoid `any`

File: lib/gui/modules/child-writer.ts [hunk 5]
```diff
@@ -131,12 +131,23 @@ async function writeAndValidate(
 	return result;
 }
 
+type Source =
```
Comment (zvin) [inline lib/gui/modules/child-writer.ts:134]: `Source` and `SourceOptions` are defined several times.

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 6]
```diff
@@ -73,36 +77,158 @@ function getState() {
...
+		}
+	}
+`;
+
+type Source = typeof sourceDestination.File | typeof sourceDestination.Http;
+
+interface SourceOptions {
+	imagePath: string;
+	SourceType: Source;
+	sourceParams?: any[];
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:179]: Please avoid `any`

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 7]
```diff
@@ -73,36 +77,158 @@ function getState() {
...
-export class ImageSelector extends React.Component<
-	ImageSelectorProps,
-	ImageSelectorState
+export class SourceSelector extends React.Component<
+	SourceSelectorProps,
+	SourceSelectorState
 > {
 	private unsubscribe: () => void;
+	private afterSelected: SourceSelectorProps['afterSelected'];
+	public flows: Flow[];
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:202]: No need for this

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 8]
```diff
@@ -363,15 +518,9 @@ export class ImageSelector extends React.Component<
...
-								<StepButton onClick={this.openImageSelector}>
-									Select image
-								</StepButton>
-								<Footer>
-									{mainSupportedExtensions.join(', ')}, and{' '}
-									<Underline tooltip={extraSupportedExtensions.join(', ')}>
-										many more
-									</Underline>
-								</Footer>
+								{_.map(this.flows, flow => {
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:521]: No need to define `this.flows` and use `_.map`, just use `<FlowSelector />` two times.

File: lib/gui/app/components/svg-icon/svg-icon.tsx [hunk 9]
```diff
@@ -95,31 +95,36 @@ export class SVGIcon extends React.Component<SVGIconProps> {
...
-				// This means the path to the icon should be
-				// relative to *this directory*.
-				// TODO: There might be a way to compute the path
-				// relatively to the `index.html`.
-				const imagePath = path.join(baseDirectory, 'assets', relativePath);
-
-				const contents = _.attempt(() => {
-					return fs.readFileSync(imagePath, {
-						encoding: 'utf8',
+				if (_.startsWith(relativePath, 'http')) {
```
Comment (zvin) [inline lib/gui/app/components/svg-icon/svg-icon.tsx:98]: Why this changes? SVGIcon is not used in this pull request.

File: lib/gui/app/components/svg-icon/svg-icon.tsx [hunk 10]
```diff
@@ -95,31 +95,36 @@ export class SVGIcon extends React.Component<SVGIconProps> {
...
+					svgData = relativePath;
+					return true;
+				} else {
```
Comment (zvin) [inline lib/gui/app/components/svg-icon/svg-icon.tsx:101]: No need for this `else`, the `if` has a `return`.

File: lib/gui/app/pages/main/Flash.tsx [hunk 11]
```diff
@@ -71,7 +71,10 @@ const getErrorMessageFromCode = (errorCode: string) => {
 	return '';
 };
 
-async function flashImageToDrive(goToSuccess: () => void): Promise<string> {
+async function flashImageToDrive(
+	goToSuccess: () => void,
+	source: any,
```
Comment (zvin) [inline lib/gui/app/pages/main/Flash.tsx:76]: Please avoid `any`

File: lib/gui/app/pages/main/Flash.tsx [hunk 12]
```diff
@@ -160,7 +163,11 @@ const formatSeconds = (totalSeconds: number) => {
 	return `${minutes}m${seconds}s`;
 };
 
-export const Flash = ({ shouldFlashStepBeDisabled, goToSuccess }: any) => {
+export const Flash = ({
+	shouldFlashStepBeDisabled,
+	goToSuccess,
+	source,
+}: any) => {
```
Comment (zvin) [inline lib/gui/app/pages/main/Flash.tsx:170]: Please avoid `any`

File: lib/gui/app/pages/main/Flash.tsx [hunk 13]
```diff
@@ -299,7 +306,11 @@ export const Flash = ({ shouldFlashStepBeDisabled, goToSuccess }: any) => {
 					done={() => handleFlashErrorResponse(true)}
 					action={'Retry'}
 				>
-					<Txt>{errorMessage}</Txt>
+					<Txt>
+						{_.map(errorMessage.split('\n'), (message, key) => (
+							<p key={`${message}-${key}`}>{message}</p>
```
Comment (zvin) [inline lib/gui/app/pages/main/Flash.tsx:311]: `key={key}` should be enough.

File: lib/gui/app/styled-components.tsx [hunk 14]
```diff
@@ -55,30 +57,45 @@ export const ThemedProvider = (props: any) => (
 	<Provider theme={theme} {...props}></Provider>
 );
 
-export const BaseButton = styled(Button)`
-	height: 48px;
+export const Icon = styled(props => <Button plain {...props} />)`
```
Comment (zvin) [inline lib/gui/app/styled-components.tsx:60]: `IconButton` would be a better name.

File: lib/gui/app/styled-components.tsx [hunk 15]
```diff
@@ -55,30 +57,45 @@ export const ThemedProvider = (props: any) => (
...
+		> svg {
+			font-size: 1em;
+		}
+	}
 `;
 
-export const StepButton = (props: any) => (
-	<BaseButton primary {...props}></BaseButton>
-);
+export const StepButton = styled((props: any) => <Button {...props}></Button>)`
```
Comment (zvin) [inline lib/gui/app/styled-components.tsx:73]: Please avoid `any`.

File: lib/gui/app/components/source-selector/source-selector.tsx (hunk 1 above)
Comment (thundron) [reply lib/gui/app/components/source-selector/source-selector.tsx:89]: I couldn't find a way around the jsonStorage.get method (promisified) returning `unknown` as a type
As for the default list - it won't break, the default for that method is an empty object

File: lib/gui/app/pages/main/Flash.tsx (hunk 13 above)
Comment (thundron) [reply lib/gui/app/pages/main/Flash.tsx:311]: It gave me warnings using just one or the other, but admit it was in the early stages and things have changed. I'll try that

File: lib/gui/app/components/svg-icon/svg-icon.tsx (hunk 9 above)
Comment (thundron) [reply lib/gui/app/components/svg-icon/svg-icon.tsx:98]: Leftover, will remove it

File: lib/gui/app/components/source-selector/source-selector.tsx (hunk 8 above)
Comment (thundron) [reply lib/gui/app/components/source-selector/source-selector.tsx:521]: There's no need to but it will look better with all the other flows implemented which we'll have relatively soon, and it costs nothing while reducing the clutter

File: lib/gui/app/components/source-selector/source-selector.tsx (hunk 1 above)
Comment (zvin) [reply lib/gui/app/components/source-selector/source-selector.tsx:89]: It most definitely breaks.
```
/home/alexis/dev/resin.io/etcher/node_modules/resin-corvus/src/resin-corvus.js:233 {"stack":"TypeError: recentImages is not iterable\n    at done (file:///home/alexis/dev/resin.io/etcher/generated/gui.js:4610:61)\n    at HTMLUnknownElement.callCallback (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:336:14)\n    at Object.invokeGuardedCallbackDev (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:385:16)\n    at invokeGuardedCallback (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:440:31)\n    at invokeGuardedCallbackAndCatchFirstError (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:454:25)\n    at executeDispatch (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:584:3)\n    at executeDispatchesInOrder (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:606:7)\n    at executeDispatchesAndRelease (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:713:5)\n    at executeDispatchesAndReleaseTopLevel (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:722:10)\n    at forEachAccumulated (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:694:8)\n    at runEventsInBatch (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:739:3)\n    at runExtractedPluginEventsInBatch (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:880:3)\n    at handleTopLevel (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:5807:5)\n    at batchedEventUpdates$1 (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:24368:12)\n    at batchedEventUpdates (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:1419:12)\n    at dispatchEventForPluginEventSystem (/home/alexis/dev/resin.io/etcher/node_modules/react-dom/cjs/react-dom.development.js:5898:5)","message":"recentImages is not iterable"}
```

Comment (zvin) [reply lib/gui/app/components/source-selector/source-selector.tsx:89]: Capture d’écran de 2020-04-02 16-49-01

Comment (zvin) [reply lib/gui/app/components/source-selector/source-selector.tsx:89]: For the return type, you'll need to add a typings file for electron-json-storage in the `typings` folder.

File: lib/gui/app/components/source-selector/source-selector.tsx (hunk 3 above)
Comment (thundron) [reply lib/gui/app/components/source-selector/source-selector.tsx:91]: It relies on a method that has an error parameter in the callback and rightly so because it handles opening a file, and things might go wrong.
You reminded me though that we need a warning to pop-up instead of just logging the error, as the error won't probably trigger the dialogue by itself, given that it comes from a callback (so it has to be manually handled anyway)

File: lib/gui/app/components/source-selector/source-selector.tsx (hunk 2 above)
Comment (thundron) [reply lib/gui/app/components/source-selector/source-selector.tsx:101]: It won't be, the default returned value is an empty object

Comment (thundron) [reply lib/gui/app/components/source-selector/source-selector.tsx:101]: In case it fails, it still is `[]` with React's `useState`

Comment (zvin) [reply lib/gui/app/components/source-selector/source-selector.tsx:101]: See the screenshot above

File: lib/gui/app/components/source-selector/source-selector.tsx (hunk 3 above)
Comment (zvin) [reply lib/gui/app/components/source-selector/source-selector.tsx:91]: For now, when there's an error, nothing happens, the url prompt just stays there as if no button was clicked.

File: lib/gui/app/components/source-selector/source-selector.tsx (hunk 8 above)
Comment (zvin) [reply lib/gui/app/components/source-selector/source-selector.tsx:521]: It won't look better.
You're defining a public attribute that is only used in one place in this class.

File: lib/gui/app/components/source-selector/source-selector.tsx (hunk 1 above)
Comment (thundron) [reply lib/gui/app/components/source-selector/source-selector.tsx:89]: I was able to reproduce, it's because of the default object returned. Adding a check there

Comment (thundron) [reply lib/gui/app/components/source-selector/source-selector.tsx:89]: For the return type: there are already the official typings, that's what `promisify` uses. I'll just ditch that with a custom promise, I'd rather have correct typings

Comment (zvin) [reply lib/gui/app/components/source-selector/source-selector.tsx:89]: right, I haven't seen @types/electron-json-storag

File: lib/gui/app/pages/main/Flash.tsx (hunk 12 above)
Comment (thundron) [reply lib/gui/app/pages/main/Flash.tsx:171]: this wasn't changed, but sure, pushing the fix now

**Comment by thundron** [comment]
@balena-ci retest

File: lib/gui/modules/child-writer.ts [hunk 16]
```diff
@@ -95,6 +97,7 @@ async function writeAndValidate(
 	onProgress: sdk.multiWrite.OnProgressFunction,
 	onFail: sdk.multiWrite.OnFailFunction,
 ): Promise<WriteResult> {
+	console.log('source', source);
```
Comment (zvin) [inline lib/gui/modules/child-writer.ts:100]: leftover `console.log`

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 17]
```diff
@@ -35,15 +38,50 @@ import { replaceWindowsNetworkDriveLetter } from '../../os/windows-network-drive
...
+			jsonStorageCb.get(key, (err, value) => {
+				if (err) {
+					reject(err);
+					throw err;
+				}
+				resolve(value);
+				return value;
+			});
+		});
+	},
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:61]: Throwing or returning in a `new Promise((resolve, reject) => { ... })` makes no sense.
Use `util.promisify`:
```typescript
const jsonStorage = {
    get: promisify(jsonStorageCb.get),
    set: promisify(jsonStorageCb.set),
}
```

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 18]
```diff
@@ -73,36 +101,145 @@ function getState() {
...
+	const [imageURL, setImageURL] = React.useState('');
+	const [recentImages, setRecentImages]: [
+		string[],
+		(value: React.SetStateAction<string[]>) => void,
+	] = React.useState([]);
+	React.useEffect(() => {
+		const fetchRecentUrlImages = async () => {
+			const recentUrlImages: string[] = await getRecentUrlImages();
+			if (!Array.isArray(recentUrlImages)) {
+				setRecentImages([]);
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:114]: The array check and default empty list value should be done in `getRecentUrlImages`

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 19]
```diff
@@ -35,15 +38,50 @@ import { replaceWindowsNetworkDriveLetter } from '../../os/windows-network-drive
...
+				}
+				resolve(value);
+				return value;
+			});
+		});
+	},
+};
+
+const getRecentUrlImages = () =>
+	jsonStorage.get('recentUrlImages') as Promise<string[]>;
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:77]: Please make `'recentUrlImages'` a constant and add an equivalent `setRecentUrlImages` function using it.
Check that `recentUrlImages` is an array and return a default empty array in `getRecentUrlImages`

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 20]
```diff
@@ -220,12 +357,18 @@ export class ImageSelector extends React.Component<
 			return;
 		}
 
-		const source = new sdk.sourceDestination.File({
-			path: imagePath,
-		});
+		let source;
+		if (SourceType.name === sourceDestination.File.name) {
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:361]: `(SourceType === sourceDestination.File)` should work

File: lib/gui/app/components/source-selector/source-selector.tsx (hunk 17 above)
Comment (zvin) [reply lib/gui/app/components/source-selector/source-selector.tsx:61]: Actually just use `localStorage.setItem('recentUrlImages', JSON.stringify(yourData))` and `JSON.load(localStorage.getItem('recentUrlImages'))` and remove electron-json-storage completely

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 21]
```diff
@@ -35,15 +37,39 @@ import { replaceWindowsNetworkDriveLetter } from '../../os/windows-network-drive
...
 	StepSelection,
-	Underline,
 } from '../../styled-components';
+import { colors } from '../../theme';
 import { middleEllipsis } from '../../utils/middle-ellipsis';
 import { SVGIcon } from '../svg-icon/svg-icon';
 
+const recentUrlImagesKey = 'recentUrlImages';
+
+function normalizeRecentUrlImages(urls: string[]) {
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:50]: Please add the array check, as we don't know what we might get from localStorage.
```typescript
function normalizeRecentUrlImages(urls: any): string[] {
	if (!Array.isArray(urls)) {
		urls = [];
	}
	return _.takeRight(urls.filter(_.isString), 5);
}
```

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 22]
```diff
@@ -35,15 +37,39 @@ import { replaceWindowsNetworkDriveLetter } from '../../os/windows-network-drive
...
+	return _.takeRight(urls.filter(_.isString), 5);
+}
+
+function getRecentUrlImages(): string[] {
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:54]: Actually to be 100% safe:
```typescript
function getRecentUrlImages(): string[] {
	let urls = [];
	try {
		urls = JSON.parse(localStorage.getItem(recentUrlImagesKey) || '[]');
	} catch {
		// noop
	}
	return normalizeRecentUrlImages(urls);
}
```

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 23]
```diff
@@ -73,36 +89,138 @@ function getState() {
...
+			const recentUrlImages: string[] = await getRecentUrlImages();
+			setRecentImages(recentUrlImages);
+		};
+		fetchRecentUrlImages();
+	}, []);
+	return (
+		<Modal
+			done={async () => {
+				const sanitizedRecentUrls = _.uniq(
+					_.reject([...recentImages, imageURL], _.isEmpty),
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:109]: Move this logic to `normalizeRecentUrlImages`

```typescript
const sanitizedRecentUrls = normalizeRecentUrlImages([...recentImages, imageURL])
```


```typescript
function normalizeRecentUrlImages(urls: any): string[] {
	if (!Array.isArray(urls)) {
		urls = [];
	}
	return _.chain(urls)
		.filter(_.isString)
		.reject(_.isEmpty)
		.uniq()
		.takeRight(5)
		.value();
}

```

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 24]
```diff
@@ -220,12 +338,18 @@ export class ImageSelector extends React.Component<
...
+				path: imagePath,
+			});
+		} else {
+			source = new sourceDestination.Http(imagePath);
+		}
+
 		try {
 			const innerSource = await source.getInnerSource();
-			const metadata = (await innerSource.getMetadata()) as sdk.sourceDestination.Metadata & {
+			const metadata = (await innerSource.getMetadata()) as sourceDestination.Metadata & {
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:352]: We need the UI to show that we are loading the metadata, maybe don't close the URL selector until we have the metadata.

Comment (thundron) [reply lib/gui/app/components/source-selector/source-selector.tsx:370]: Made it so that the button is disabled and it closes when the data's done loading

File: lib/gui/app/pages/main/MainPage.tsx [hunk 25]
```diff
@@ -188,12 +197,17 @@ export class MainPage extends React.Component<
...
-					<div
+					<Flex
 						className="page-main row around-xs"
 						style={{ margin: '110px 50px' }}
 					>
 						<div className="col-xs">
-							<ImageSelector flashing={this.state.isFlashing} />
+							<SourceSelector
+								flashing={this.state.isFlashing}
+								afterSelected={(source: SourceOptions) =>
```
Comment (zvin) [inline lib/gui/app/pages/main/MainPage.tsx:207]: Please throw an error if we're using `Http` and `source.imagePath` does not start with `http://` or `https://`
Something like "Only http:// and https:// protocols are supported"

Comment (thundron) [reply lib/gui/app/pages/main/MainPage.tsx:207]: Added after checking for supported image formats too

File: lib/gui/app/components/source-selector/source-selector.tsx [hunk 26]
```diff
@@ -220,12 +356,28 @@ export class ImageSelector extends React.Component<
...
-		const source = new sdk.sourceDestination.File({
-			path: imagePath,
-		});
+		let source;
+		if (SourceType === sourceDestination.File) {
+			source = new sourceDestination.File({
+				path: imagePath,
+			});
+		} else {
+			if (!_.startsWith(imagePath, 'http')) {
```
Comment (zvin) [inline lib/gui/app/components/source-selector/source-selector.tsx:365]: Please explicitly check for 'http://' or 'https://'.
This code won't throw for `httpwhatever://invalid-url`

File: lib/shared/messages.ts [hunk 27]
```diff
@@ -191,4 +191,8 @@ export const error = {
 			'Please try again, and contact the Etcher team if the problem persists.',
 		].join(' ');
 	},
+
+	unsupportedProtocol: () => {
+		return 'Please insert only URLs that use HTTP or HTTPS.';
```
Comment (zvin) [inline lib/shared/messages.ts:196]: "Only http:// and https:// URLs are supported"

File: lib/gui/app/components/source-selector/source-selector.tsx (hunk 26 above)
Comment (thundron) [reply lib/gui/app/components/source-selector/source-selector.tsx:365]: It will throw as an invalid image so this code is actually redundant for the purpose of a validity check.
I'm just adding it as a way to instruct users, because it throws anyway
